enterprise/backend/src/metabase_enterprise/semantic_search/embedding.clj

```diff
@@ -83,159 +81,140 @@
-        (throw e))))
+(defmethod get-embedding        "openai" [{:keys [model-name]} text] (openai-get-embedding model-name text))
+(defmethod get-embeddings-batch "openai" [{:keys [model-name]} text] (openai-get-embeddings-batch model-name text))
+(defmethod pull-model           "openai" [_] (log/info "OpenAI provider does not require pulling a model"))
 
-  (-model-dimensions [_ model]
-    (get ollama-supported-models model)))
+;;;; Global embedding model
 
-(defrecord OpenAIProvider []
-  EmbeddingProvider
-  (-get-embedding [this text]
-    (try
-      (log/debug "Generating OpenAI embedding for text of length:" (count text))
-      (first (-get-embeddings-batch this [text]))
-      (catch Exception e
-        (log/error e "Failed to generate OpenAI embedding for text of length:" (count text))
-        (throw e))))
-
-  (-get-embeddings-batch [_ texts]
-    (let [api-key (semantic-settings/openai-api-key)
-          model (get-model)
-          endpoint "https://api.openai.com/v1/embeddings"]
-      (when-not api-key
-        (throw (ex-info "OpenAI API key not configured" {:setting "ee-openai-api-key"})))
-      (try
-        (log/debug "Generating" (count texts) "OpenAI embeddings in batch")
-        (-> (http/post endpoint
-                       {:headers {"Content-Type" "application/json"
-                                  "Authorization" (str "Bearer " api-key)}
-                        :body    (json/encode {:model model
-                                               :input texts})})
-            :body
-            (json/decode true)
-            :data
-            (->> (map :embedding)
-                 vec))
-        (catch Exception e
-          (log/error e "Failed to generate OpenAI embeddings for batch of" (count texts) "texts"
-                     "with token count:" (count-tokens-batch texts))
-          (throw e)))))
-
-  (-pull-model [_]
-    (log/info "OpenAI provider does not require pulling a model"))
-
-  (-model-dimensions [_ model]
-    (get openai-supported-models model)))
-
-(def ^:private providers
-  "Registry of available embedding providers."
-  {:ollama (->OllamaProvider)
-   :openai (->OpenAIProvider)})
-
-(defn get-provider
-  "Get the configured embedding provider according to the ee-embedding-provider setting"
-  []
-  (let [provider-key (keyword (semantic-settings/ee-embedding-provider))]
-    (if-let [provider (get providers provider-key)]
-      provider
-      (throw (ex-info (str "Unknown embedding provider: " provider-key)
-                      {:provider provider-key
-                       :available-providers (keys providers)})))))
-
-(defn get-embedding
-  "Generate a single embedding using the configured provider. Prefer using `process-embeddings-streaming` for bulk index population."
-  [text]
-  (-get-embedding (get-provider) text))
-
-(defn pull-model
-  "Pull/download the model using the configured provider."
+(def ^:dynamic ^:redef *active-model*
```

Comment:
What's the purpose of the `^:redef` metadata? Not super familiar but my understanding is it [opts it out of direct linking](https://clojure.org/reference/compilation#directlinking) which I don't believe we're using. Is that relevant here?

Reply:
Oops, yes its a mistake this was left in.

FWIW At one point I was redefining it. Even if direct linking is not used today I find it good practice to mark vars you expect to bang on with `:redef` if only as a signal to readers that we expect to redefine it. 

i.e instead of the atom in db.clj we could have a var and .doReset it to avoid the extra deref, as we are not really using the atom. In this case I would use `:redef` to emphasise the 'globally mutable variable, not only at the repl' flavor of this var.

Reply:
11aeb2346c5cc26e5237753de1c7cdd658ec8a51